lnrpc/invoicesrpc/invoices.proto

```diff
@@ -41,6 +41,13 @@ service Invoices {
     using either its payment hash, payment address, or set ID.
     */
     rpc LookupInvoiceV2 (LookupInvoiceMsg) returns (lnrpc.Invoice);
+
+    /*
+    GetOrGeneratePreimageHash returns the preimage and its hash. If the 
+    preimage is present in the request, generate the hash for it. Otherwise, 
+    create a new preimage and return its hash. 
+    */
+    rpc GetOrGeneratePreimageHash (GetPreimageHashRequest) returns (GetPreimageHashResponse);
 }
```

Comment:
I think the intention of this issue was to explicitly change the [AddHoldInvoice](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/invoicesrpc/invoices.proto#L31) so  maybe check if the hash is `nil` and if so, let lnd add the preimage + hash  and return it in the `AddHoldInvoiceResp` adding the preimage to the response

Reply:
and on the `lncli addholdinvoice` side, you need to allow a nil hash ...

Reply:
> Thank you for taking care of this issue @joyqvq
> 
> I left a couple of comments + some cuestions in the issue itself.
> 
> For testing this we would have to add a new itest similar to `testHoldInvoicePersistence`.

dumb q, is this test ran with `make itest`? from the log output im not very convinced the test is ran, but it does belong to the itest suite right?

```
make itest
 Building itest btcd and lnd.
CGO_ENABLED=0 go build -v -tags="rpctest" -o lntest/itest/btcd-itest -ldflags " -X github.com/lightningnetwork/lnd/build.Commit=kvdb/v1.4.1-35-g1a53edae8-dirty" github.com/btcsuite/btcd
CGO_ENABLED=0 go build -v -tags="dev autopilotrpc chainrpc invoicesrpc neutrinorpc peersrpc routerrpc signrpc verrpc walletrpc watchtowerrpc wtclientrpc rpctest btcd" -o lntest/itest/lnd-itest -ldflags " -X github.com/lightningnetwork/lnd/build.Commit=kvdb/v1.4.1-35-g1a53edae8-dirty" github.com/lightningnetwork/lnd/cmd/lnd
 Building itest binary for btcd backend.
CGO_ENABLED=0 go test -v ./lntest/itest -tags="dev autopilotrpc chainrpc invoicesrpc neutrinorpc peersrpc routerrpc signrpc verrpc walletrpc watchtowerrpc wtclientrpc rpctest btcd" -c -o lntest/itest/itest.test
 Running integration tests with btcd backend.
rm -rf lntest/itest/*.log lntest/itest/.logs-*; date
Sun Feb 12 22:32:24 EST 2023
EXEC_SUFFIX= scripts/itest_part.sh 0 1  -test.timeout=120m
/Users/joy/refs/lnd/lntest/itest/itest.test -test.v -test.timeout=120m -logoutput -goroutinedump -logdir=.logs-tranche0 -lndexec=/Users/joy/refs/lnd/lntest/itest/lnd-itest -btcdexec=/Users/joy/refs/lnd/lntest/itest/btcd-itest -splittranches=1 -runtranche=0
=== RUN   TestLightningNetworkDaemon
=== RUN   TestLightningNetworkDaemon/tranche00/01-of-1/btcd/async_bidirectional_payments
--- PASS: TestLightningNetworkDaemon (87.66s)
    --- PASS: TestLightningNetworkDaemon/tranche00/01-of-1/btcd/async_bidirectional_payments (86.84s)
=== RUN   TestLightningNetworkDaemonTemp
    temp_lnd_test.go:59: Running old tests, new tests are skipped
--- SKIP: TestLightningNetworkDaemonTemp (0.00s)
PASS
```

Reply:
another beginner q, where can I find my built lnd and lncli binary and add it to bash? i think i installed them (following: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) but not sure how to use
```make install                                                                                                
Installing lnd and lncli.
go install -v -tags="" -ldflags=" -s -w -buildid= -X github.com/lightningnetwork/lnd/build.Commit=kvdb/v1.4.1-35-g89a87ad5b" github.com/lightningnetwork/lnd/cmd/lnd
go install -v -tags="" -ldflags=" -s -w -buildid= -X github.com/lightningnetwork/lnd/build.Commit=kvdb/v1.4.1-35-g89a87ad5b" github.com/lightningnetwork/lnd/cmd/lncli

lncli                                                                                                                 
zsh: command not found: lncli
```

Reply:
question1: `make itest icase=hold_invoice_sender_persistence temptest=1`(you need to run the new tests with temptest=1)

2) make sure the GOPATH is part of your PATH variable

Reply:
> question1: `make itest icase=hold_invoice_sender_persistence temptest=1`(you need to run the new tests with temptest=1)
> 
> 
> 
> question2:  make sure the GOPATH is part of your PATH variable

for make itest, is there a way to enable print or anyway to debug? i added some print in my code but it doesn't seem to be working.

Reply:
the itest creates different lnd instances (e.g. Alice or Bob) which log their output to temporary files you will find it in the temptest folder : `lnd/lntest/itest/.logs-tranche0`

Reply:
including the bitcoin backend

Reply:
logs you write in the itest should be show though. Only the logs written in the lnd daemon will not be shown but only in the aforementioned log-files.

Reply:
fixed tests - PTAL. 

another question, i see there is a merge conflict, whats the best way to resolve? (i usually `git rebase master`, should i pull from latest master instead?)

Reply:
hm thats what i did,`git pull master` and `git checkout my-branch` and `git rebase master`, it doesnt seem to pick up the merge conflict. is there a guide for this?